### ISSUE TITLE
Any plans to support pcap_setdirection()?

### ISSUE THREAD

**Issue by pstavirs** [nmap/npcap#248]
Similar to WinPcap, Npcap also doesn't seem to support pcap_setdirection.

Any plans to support?

Using Npcap 0.9990

**Comment by guyharris** [comment]
Is there a way to get an "incoming vs. outgoing" indication from NDIS?

**Comment by dmiller-nmap** [comment]
For NDIS filter drivers like Npcap, direction is determined by whether the packet is in a "send indication" or a "receive indication." NDIS calls separate callback functions for each of these, and Npcap ends up processing packets in the same `NPF_DoTap` function. We could potentially support a capture option or mode for this, but we'll have to figure out priority.

@pstavirs You should definitely update to Npcap 1.00.

**Comment by guyharris** [comment]
> For NDIS filter drivers like Npcap, direction is determined by whether the packet is in a "send indication" or a "receive indication." NDIS calls separate callback functions for each of these, and Npcap ends up processing packets in the same `NPF_DoTap` function. We could potentially support a capture option or mode for this

Yes - you'd have a per-instance mode settable with an ioctl, with options "process all packets", "process received packets only", and "process sent packets only" (or "ignore no packets", "ignore sent packets", and "ignore received packets" - that could be implemented as two "ignore" flags).

> @pstavirs You should definitely update to Npcap 1.00.

You won't get `pcap_setdirection()`, but you *will* get bug fixes, performance improvements, etc..

**Comment by guyharris** [comment]
Distinguishing between received and sent could also be useful in the future; at some point I'd like to have a new capture API, delivering pcapng blocks, which would allow the packet direction to be supplied as the epb_flags option.  `struct bpf_hdr` has a `bh_hdrlen` option - one (possibly unintended) side-effect of that is that if readers honor that field, rather than assuming the structures provided by a read have a fixed size, is that more information can be added to the structure after `bh_hdrlen`, without breaking binary compatibility.

(That field originally came from BPF, which added it to allow padding to be inserted before the link-layer header, so that, for example, the IP header could be put on a 4-byte boundary - but it also allows additional information to be provided.  Apple's extended BPF header is only used if you ask for it - they may have worried about programs using BPF and *not* using `bh_hdrlen`, and maybe that'd be an issue in \*BSD as well - but I'm guessing that wouldn't be an issue for NPF.)

**Comment by fyodor** [comment]
I've received some other requests for this as well.  It sounds like it wouldn't be too had to implement given that the libpcap API exists and that NDIS uses different callbacks so we can distinguish sent vs received packets that way.  The new pcapng API stuff @guyharris talks about for determining direction of received packets might be a longer term goal, but for now (once we have pcap_setdirection support), applications which care could open two handles. One for capturing packets sent by the system and one for those received.

**Comment by guyharris** [comment]
> The new pcapng API stuff @guyharris talks about for determining direction of received packets might be a longer term goal,

Yes - [pcapng format](https://pcapng.github.io/pcapng/draft-ietf-opsawg-pcapng.html) includes an option for the [Enhanced Packet Block](https://pcapng.github.io/pcapng/draft-ietf-opsawg-pcapng.html#name-enhanced-packet-block) that includes [flags that indicate whether the packet was sent by the capturing host, the packet was received by the capturing host, or it's unknown whether the packet was sent or received by the capturing host](https://pcapng.github.io/pcapng/draft-ietf-opsawg-pcapng.html#name-enhanced-packet-block-flags).

An API that provides EPB-style information for each packet when capturing or reading would allow that information to be provided on platforms that provide it.

**Comment by pstavirs** [comment]
Any plans for this?

**Comment by Saigut** [comment]
Any plans?

**Comment by farcas-snps** [comment]
This is quite embarrassing. Any plans for supporting filtering by direction?

**Comment by Kollateral9** [comment]
**Transparent Layer 2 loopback test bench. Npcap 1.88**

I send raw Ethernet frames from a host to an embedded device that bridges
Ethernet <-> UART and echoes them back byte-for-byte. Verifying that the bridge
is byte-transparent *is* the test.

The capture also shows the frames I send, and because the device is transparent
the echo is byte-identical to what went out. Nothing in the frame can tell the
two apart, which rules out every content-based workaround suggested in threads
like this one:

- BPF filter options: drops my own copy and the echo along with it, so, not an option here.
- checking the FCS: stripped by the capture, and both copies would carry the
  same correct value over the same bytes anyway, so, no again.
- hashing sent frames and discarding matches: the standard bridge remedy, but
  `hash(own copy) == hash(echo)` by construction, so, no way to know if the embedded device has responded correctly.

This means that direction metadata *would* be the solution. And on Linux `AF_PACKET` hands it over
for free via `PACKET_OUTGOING`. 

On Windows I tried:

`pcap_setdirection(handle, PCAP_D_IN)` but this gives me RC= -1 with "Setting direction is not supported on this device"

So I had to fallback on the classic method of the timestamps BUT **The timing fallback is not equivalent!!!** 

I currently separate my own copy from the echo by arrival time: an echo cannot come back faster than the
frame takes to cross the serial link. That holds only while the two populations
stay apart. My own transmission reappears in the capture ~800 µs later — a
property of the capture path, essentially independent of frame size — while the
serial floor grows with the frame (3 Mbit/s UART):

| frame | serial floor | own copy at ~800 µs |
| --- | --- | --- |
| 60 B | 220 µs | past the floor — indistinguishable from an echo |
| 200 B | 687 µs | past the floor — indistinguishable from an echo |
| 260 B | 890 µs | correctly identified |
| 1514 B | 5083 µs | correctly identified |

Below roughly 240 bytes the populations overlap and no rule applied to timestamps can separate them. 
And this opens to a real issue of test coverage:  **a board that had never been flashed returned only my own frames and passed the entire suite, silently**. 
Every frame size below the crossover is **untestable** on Windows and testable on Linux, from identical application code.

That's a problem for me and for everyone would ever develop such applications on Windows.

From my point of view, with my (little) knowledge base **implements this feature looks feasible**: 

An NDIS lightweight filter receives the two paths in separate callbacks: `FilterSendNetBufferLists` and
`FilterReceiveNetBufferLists`, so the direction is known at the moment of capture and discarded before it reaches userspace. Even a partial implementation would cover this entire class of application: `PCAP_D_IN` only, Ethernet adapters only, `PCAP_D_OUT` still returning an error.

I recognise that "the information is right there" and "it is cheap to expose" are different claims, and I have no visibility into what the LWF and the WinPcap-compatibility layer make of it. If there is a structural reason this cannot be done, saying so in the issue would still help! 

People keep arriving here and reinventing the same unsound timing workarounds, mine included.

Meanwhile I am testing `pcap_open()` with `PCAP_OPENFLAG_NOCAPTURE_LOCAL` as a substitute and will report what it does on a physical adapter, in case that is a cheaper route to the same outcome for others reading this.

**Comment by Kollateral9** [comment]
I had some times for testing and no good news:

| workaround | result |
| --- | --- |
| `pcap_open()` + `PCAP_OPENFLAG_NOCAPTURE_LOCAL` | handle opens, no observable effect |
| `PacketSetLoopbackBehavior(NPF_DISABLE_LOOPBACK)` | no effect from another instance |

**Setup:** Npcap 1.88, libpcap 1.10.6, Windows 10, physical Ethernet adapter.)

**Method:** 
Five 1514-byte frames, counting the copies that come back. The baseline
shows two per frame: our own transmission at 250-880 µs and the echo from the
device at ~13.4 ms. Under both workarounds it stayed two per frame

**Note:**
Set `pcap_setmintocopy(p, 0)` on a hand-opened handle.
Npcap otherwise holds packets until 16 kB accumulate, delivery is batched, and
our own copy surfaces ~15 ms late, above the threshold we classify by, which
first looked exactly like NOCAPTURE_LOCAL working. scapy zeroes it on every
handle it opens, which is why the baseline was unaffected.

**Conclusion:**
There is currently no working way to filter capture direction on Windows,
including the two that are documented as being for exactly this. Making
`PCAP_OPENFLAG_NOCAPTURE_LOCAL` do what it says may well be cheaper than
implementing `pcap_setdirection`, since the flag is already in the API and it
would close this for every application that has reached for it.

This really needs a fix.
